packages/workflow/src/internals.ts

```diff
@@ -534,14 +542,96 @@ export class Activator implements ActivationHandler {
     );
   }
 
+  public doUpdate(activation: coresdk.workflow_activation.IDoUpdate): void {
+    const { id, name, headers, runValidator } = activation;
+    if (!id) {
+      throw new TypeError('Missing activation update id');
+    }
+    if (!name) {
+      throw new TypeError('Missing activation update name');
+    }
+    if (!this.updateHandlers.has(name)) {
+      // TODO (dan): Signal is able to handle this situation more gracefully by
+      // using this.bufferedSignals. Should something analogous exist for
+      // Update?
+      this.rejectUpdate(id, new ApplicationFailure(`Update has no handler: ${name}`));
+      return;
+    }
+    const validate = composeInterceptors(
+      this.interceptors.inbound,
+      'validateUpdate',
+      this.validateUpdateNextHandler.bind(this)
+    );
+    const execute = composeInterceptors(this.interceptors.inbound, 'handleUpdate', this.updateNextHandler.bind(this));
+
+    const makeInput = (): UpdateInput => ({
+      args: arrayFromPayloads(this.payloadConverter, activation.input),
+      name,
+      headers: headers ?? {},
+    });
+
+    // The implementation below is responsible for upholding the following
+    // guarantees:
+    //
+    // 1. The initial synchronous portion of the (async) Update handler should
+    //    be executed "as soon as possible" after the (sync) validator
+    //    completes. Ideally, there is no possibility that a different
+    //    concurrent task can be scheduled between these.
+    //
+    // 2. During validation, any error must fail the Update; during the Update
+    //    itself, Temporal errors fail the Update whereas other errors fail the
+    //    activation.
+    //
+    // 3. The handler will not see any mutations of the arguments made by the
+    //    validator.
+    //
+    // Note that there is a deliberately unhandled promise rejection below.
+    // These are caught elsewhere and fail the corresponding activation.
+
+    // Due to (1), we do not use `return await ...`.
+    // TODO (dan): Read https://v8.dev/blog/fast-async: should we use `return await` or not?
+    let accepted = false;
+    const _doUpdate = async () => {
+      if (runValidator) {
+        validate(makeInput());
+      }
+      this.acceptUpdate(id);
+      accepted = true;
+      return execute(makeInput()).then((result) => this.completeUpdate(id, result));
+    };
+    _doUpdate().catch((error) => {
+      if (!accepted || error instanceof TemporalFailure) {
+        this.rejectUpdate(id, error);
+      } else {
+        throw error;
+      }
+    });
+  }
+
+  public async updateNextHandler({ name, args }: UpdateInput): Promise<unknown> {
+    const [fn, _] = this.updateHandlers.get(name) ?? [];
+    if (fn) {
+      return await fn(...args);
+    } else {
+      throw new IllegalStateError(`No registered update handler for update: ${name}`);
+    }
```

Comment:
nit: prefer early return over nesting.

```suggestion
    if (!fn) {
      throw new IllegalStateError(`No registered update handler for update: ${name}`);
    }
    return await fn(...args);
```